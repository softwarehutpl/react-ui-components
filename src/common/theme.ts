const theme: { [key: string]: { [key: string]: { [key: string]: string } } } = {
  colors: {
    primary: {
      base: '#000000',
      light: '#ffffff',
      dark: '#a8b3bd',
    },
    secondary: {
      base: '#3690e3',
      light: '#75b0e6',
      dark: '#034d91',
    },
    error: {
      base: '#f44336',
      light: '#f56c62',
      dark: '#a1140a',
    },
    warning: {
      base: '#faa72d',
      light: '#ffc36b',
      dark: '#995b00',
    },
    info: {
      base: '#7979d4',
      light: '#9797d1',
      dark: '#4c4cc7',
    },
    success: {
      base: '#4caf50',
      light: '#88c28a',
      dark: '#124f14',
    },
    disabled: {
      base: '#d9d9d9',
      light: '#f0eded',
      dark: '#8c8989',
    },
  },
  // TODO: change it in dedicated task for that (change ease-out for example and add more options)
  transitions: {
    all: {
      fast: 'all 0.5s ease-out',
      mid: 'all 2s ease-out',
      slow: 'all 4s ease-out',
    },
    background: {
      fast: 'background-color 0.5s ease-out',
      mid: 'background-color 2s ease-out',
      slow: 'background-color 4s ease-out',
    },
    fontColor: {
      fast: 'color 0.5s ease-out',
      mid: 'color 2s ease-out',
      slow: 'color 4s ease-out',
    }
  }
};

export default theme;
