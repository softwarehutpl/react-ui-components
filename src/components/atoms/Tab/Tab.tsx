import React from 'react';
import styled, { css } from 'styled-components';
import { ORIENTATION } from '../../../common/consts/consts';

export interface WrapperProps {
  color?: string;
  fontColor?: string;
  isSelected?: boolean;
  disabled?: boolean;
  backgroundColor?: string;
  activeColor?: string;
  activeBackgroundColor?: string;
  disabledColor?: string;
  disabledBackgroudColor?: string;
  padding?: number;
  orientation?: string;
  iconClassName?: string;
};

export interface LabelProps {
  isSelected?: boolean;
};

export interface TabProps extends WrapperProps, LabelProps {
  className?: string;
  label: string;
  value?: any;
  onClick: (value: any) => void;
  icon?: any;
};

const createBorderCSSProp = (orientation: string) => {
  let borderSide;
  if(orientation === ORIENTATION.HORIZONTAL) {
    borderSide = 'bottom';
  } else {
    borderSide= 'right';
  }
  return borderSide;
};
 
export const Wrapper = styled.div<WrapperProps>`
  width: 100%;
  cursor: pointer;
  color: ${({ fontColor, theme, color, activeColor, isSelected }) => (fontColor
    || (activeColor && activeColor)
    || ((color && isSelected ) ? theme.colors[color].base : theme.colors.disabled.dark))};
  background-color: ${({ backgroundColor, theme, color, activeBackgroundColor }) => (backgroundColor 
    || ( activeBackgroundColor &&  activeBackgroundColor) 
    || ( color && theme.colors[color].base))};
  padding: ${({ padding }) => (padding && `${padding}px`)};
  font-weight: 700;
  transition: ${({ theme }) => (theme.transitions.all.fast)};

  ${({ orientation, color, isSelected, theme }) => 
    orientation && css `
      border-${createBorderCSSProp(orientation)}-style: solid;
      border-${createBorderCSSProp(orientation)}-width: ${() => (isSelected ? '2px'  : '1px')};
      border-${createBorderCSSProp(orientation)}-color: ${() => (color && (isSelected
      ? theme.colors[color].base : theme.colors.disabled.light))};
    `
  }

  ${({ disabled, disabledColor, disabledBackgroudColor }) =>
    disabled && 
      css`
        cursor: not-allowed;
        opacity: 0.4;
        color: ${() => (disabledColor && disabledColor)};
        background-color: ${() => (disabledBackgroudColor && disabledBackgroudColor)};
    `
  }
`;

export const IconWrapper = styled.div`
  display: inline-block;
  width: 24px;
  height: 24px;
  padding-right: 8px;
`;

const Tab: React.StatelessComponent<TabProps> = ({
  className,
  iconClassName,
  label,
  value,
  onClick,
  icon,
  isSelected,
  disabled,
  fontColor,
  color,
  backgroundColor,
  padding,
  activeColor,
  activeBackgroundColor,
  disabledColor,
  disabledBackgroudColor,
  orientation
}) => (
  <Wrapper
    className={className}
    onClick={() => onClick(value)}
    disabled={disabled}
    fontColor={fontColor}
    color={color}
    backgroundColor={backgroundColor}
    padding={padding}
    activeColor={activeColor}
    activeBackgroundColor={activeBackgroundColor}
    disabledColor={disabledColor}
    disabledBackgroudColor={disabledBackgroudColor}
    isSelected={isSelected}
    orientation={orientation}
  >
    <IconWrapper className={iconClassName}>{icon}</IconWrapper>
    <span>{label}</span>
  </Wrapper>
);

Tab.defaultProps = {
  color: "primary",
  orientation: 'horizontal',
  className: '',
  disabled: false,
  backgroundColor: '#fff',
  padding: 10,
};

export default Tab;
