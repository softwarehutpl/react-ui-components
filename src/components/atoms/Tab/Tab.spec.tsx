import React from 'react';
import { create, act } from 'react-test-renderer';
import 'jest-styled-components';
import Tab, { Wrapper } from './Tab';
import theme from '../../../common/theme';
import { ThemeProvider } from 'styled-components';

describe('<Tab />', () => {
  it('should render corectly', () => {
    const component = create(<Wrapper theme={theme} />);
    expect(component).toMatchSnapshot();
  });

  it('should render with custom class', () => {
    const component = create(<Wrapper className="test-class" theme={theme} />);
    const instance = component.root;
    instance.findByType('div');
    expect(instance.props.className).toEqual(expect.stringContaining('test-class'));
  });

  it('should render with styles based on theme color', () => {
    const component = create(<Wrapper color="primary" theme={theme} />).toJSON();
    expect(component).toHaveStyleRule('color', '#8c8989');
  });

  it('should render with color based on fontColor prop', () => {
    const component = create(<Wrapper fontColor="blue" theme={theme} />).toJSON();
    expect(component).toHaveStyleRule('color', 'blue');
  });

  it('should call onClick method', () => {
    const mockFn = jest.fn();
    const component = create(
      <ThemeProvider theme={theme}>
        <Tab label="tab"  onClick={mockFn} />
      </ThemeProvider>
    );
    const instance = component.root;
    const tab = instance.findByType(Tab);

    act(() =>  {
      tab.props.onClick();
    });
    expect(mockFn.mock.calls.length).toEqual(1);
  });

  it('should have proper style when disabled', () => {
    const component = create(<Wrapper color="primary" theme={theme} disabled />).toJSON();
    expect(component).toHaveStyleRule('cursor', 'not-allowed');
    expect(component).toHaveStyleRule('opacity', '0.4');
  });

  it('should have proper style when backgroundColor and color is provided', () => {
    const component = create(
      <Wrapper 
        theme={theme} 
        backgroundColor="grey"
        fontColor="pink"
      />
    ).toJSON();
    expect(component).toHaveStyleRule('background-color', 'grey');
    expect(component).toHaveStyleRule('color', 'pink');
  });

  it('should have proper style when active backgroundColor and active color is provided', () => {
    const component = create(
      <Wrapper 
        theme={theme} 
        activeBackgroundColor="blue"
        activeColor="green"
      />
    ).toJSON();
    expect(component).toHaveStyleRule('background-color', 'blue');
    expect(component).toHaveStyleRule('color', 'green');;
  });

  it('should have proper style when disabled backgroundColor and disabled color is provided', () => {
    const component = create(
      <Wrapper 
        theme={theme} 
        disabledBackgroudColor="black"
        disabledColor="grey"
        disabled
      />
    ).toJSON();
    expect(component).toHaveStyleRule('background-color', 'black');
    expect(component).toHaveStyleRule('color', 'grey');
  });

  it('should reneder propery style when padding is provided', () => {
    const component = create(
      <Wrapper 
        theme={theme} 
        padding={10}
      />
    ).toJSON();
    expect(component).toHaveStyleRule('padding', '10px');
  });
});
