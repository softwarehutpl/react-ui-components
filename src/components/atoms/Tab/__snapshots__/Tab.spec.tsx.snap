// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tab /> should render corectly 1`] = `
.c0 {
  width: 100%;
  cursor: pointer;
  color: #8c8989;
  font-weight: 700;
  -webkit-transition: all 0.5s ease-out;
  transition: all 0.5s ease-out;
}

<div
  className="c0"
/>
`;
